### PR TITLE
interfaces: add wireless-control

### PR DESCRIPTION
**Pull request by lpotter**

### CODE REVIEW

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

File: snap/implicit.go [hunk 1]
```diff
@@ -42,6 +42,7 @@ var implicitSlots = []string{
 	"system-observe",
 	"timeserver-control",
 	"timezone-control",
+    "wireless-control",
```
Comment (niemeyer) [inline snap/implicit.go]: Indentation seems off here.

**Comment by niemeyer** [comment]
Is this wireless-control or wifi-control?  Do we intend to give it privilege over non-wifi technology?

Seems straightforward code-wise.

Keeping security details for @jdstrand.

**Comment by morphis** [comment]
@niemeyer: This should be wireless-control to keep it in line with other things in the Linux world like the wireless-tools package in debian.

File: interfaces/builtin/wireless-control.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+
+const  wirelessControlConnectedPlugAppArmor = `
+# Description: Can configure wireless networking. This is restricted because it gives
+# wide, privileged access to wireless networking and should only be used with trusted
+# apps.
+# Usage: reserved
+
+capability net_admin,
+capability setgid,
```
Comment (morphis) [inline interfaces/builtin/wireless-control.go]: For what do we need setgid? This is not related to wireless management. If the snap needs it that is out of the scope for this interface.

File: interfaces/builtin/wireless-control.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+# wireless tools
+/{,usr/}{,s}bin/iw ixr,
+/{,usr/}{,s}bin/iwconfig ixr,
+/{,usr/}{,s}bin/iwevent ixr,
+/{,usr/}{,s}bin/iwgetid ixr,
+/{,usr/}{,s}bin/iiwlist ixr,
+/{,usr/}{,s}bin/iwpriv ixr,
+/{,usr/}{,s}bin/iwspy ixr,
+
+/{,usr/}{,s}bin/dnsmasq ixr,
```
Comment (morphis) [inline interfaces/builtin/wireless-control.go]: Why do we have dnsmasq and hostapd here? Those two are not directly related to this interface. If a snap wants access to dnsmasq/hostapd from the OS snap that needs to be done differently.

File: interfaces/builtin/wireless-control.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+/{,usr/}{,s}bin/hostapd ixr,
+
+/dev/rfkill r,
+/sys/class/net/ r,
+/sys/class/net/** r,
+/bin/sync ixr,
```
Comment (morphis) [inline interfaces/builtin/wireless-control.go]: /bin/sync isn't really related to a "wireless-control" interface. Drop it.

File: interfaces/builtin/wireless-control.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+#include <abstractions/nameservice>
+
+# DBus accesses
+#include <abstractions/dbus-strict>
+
+# Allow access to wpa-supplicant for managing WiFi networks
+dbus (receive, send)
```
Comment (morphis) [inline interfaces/builtin/wireless-control.go]: For which reason do we have dbus rules in here? I don't see any reasons for this interface caring them. Its just about enabling raw wireless management and nothing else.

File: interfaces/builtin/wireless-control.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+send
+sendmmsg
+sendmsg
+sendto
+setsockopt
+shutdown
+socketpair
+socket
+
+chown
```
Comment (morphis) [inline interfaces/builtin/wireless-control.go]: Those shouldn't we here. They might be relevant for the sw-access-point snap but not for a general wireless-control interface. Drop those. We need to handle those differently.

**Comment by morphis** [comment]
@niemeyer, @jdstrand @zyga The intent of this interface is to give snaps access to wireless management resources in the system but not more. There are currently things in the interface (see my comments) we need to drop to get to this minimized interface version.

File: interfaces/builtin/wireless-control.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+@{PROC}/sys/net/nf_conntrack_max rw,
+
+# wireless tools
+/{,usr/}{,s}bin/iw ixr,
+/{,usr/}{,s}bin/iwconfig ixr,
+/{,usr/}{,s}bin/iwevent ixr,
+/{,usr/}{,s}bin/iwgetid ixr,
+/{,usr/}{,s}bin/iiwlist ixr,
+/{,usr/}{,s}bin/iwpriv ixr,
+/{,usr/}{,s}bin/iwspy ixr,
```
Comment (niemeyer) [inline interfaces/builtin/wireless-control.go]: None of those seem to be in the core snap today.

Also, please note we have a network-manager interface today.

Given that and @morphis' comments, can we get some more context for the role of this interface and how it's being tested?

Comment (lpotter) [inline interfaces/builtin/wireless-control.go]: Not sure about auto testing, but I made a simple wireless-tools snap that I am using for development. As well, this is needed for a hostspot snap also being worked on.

File: interfaces/builtin/wireless-control.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+# Allow protocols except those that we blacklist in
+# /etc/modprobe.d/blacklist-rare-network.conf
+network appletalk,
+network bridge,
+network inet,
+network inet6,
+network ipxa,
+network packet,
+network pppox,
```
Comment (zyga) [inline interfaces/builtin/wireless-control.go]: Are all of those required?

File: interfaces/builtin/wireless-control.go (hunk 7 above)
Comment (morphis) [inline interfaces/builtin/wireless-control.go]: @lpotter As @niemeyer said, none of these things is in the core OS snap so putting in a rule like "/{,usr/}{,s}bin/iw ixr" doesn't make sense as it doesn't apply to anything. If a snap needs those tools it has to include them on its own and then is already allowed to execute those binaries as they are in its $SNAP directory.

If the hotspot snap is relying on those tools it has to include them.

File: interfaces/builtin/wireless-control.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+recvmsg
+send
+sendmmsg
+sendmsg
+sendto
+setsockopt
+shutdown
+socketpair
+socket
+
```
Comment (morphis) [inline interfaces/builtin/wireless-control.go]: Drop the two empty lines here.

**Comment by jdstrand** [comment]
I'll comment top-level and question why this is a separate interface from network-manager and/or network-control. It seems to make the most sense to add all of this to network-control instead. There is a ton of overlap otherwise and it seems like wanting wireless-control without network-control is being fine-grained for fine-grained sake and practically speaking isn't going to give the kinds of separation that would be desired. I suggest adding to network-control:
```
# for wireless networking
/dev/wl* rw,
network netlink raw,
```

I'll continue my review with this perspective in mind.

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

File: interfaces/builtin/wireless-control.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+/{,usr/}{,s}bin/iwspy ixr,
+`
+
+const wirelessControlConnectedPlugSecComp = `
+# Description: Can configure wireless networking. This is restricted because it gives
+# wide, privileged access to wireless networking and should only be used with trusted
+# apps.
+# Usage: reserved
+
+accept
```
Comment (jdstrand) [inline interfaces/builtin/wireless-control.go]: I agree with @morphis - please drop all of these seccomp calls.

File: interfaces/builtin/wireless-control.go (hunk 7 above)
Comment (jdstrand) [inline interfaces/builtin/wireless-control.go]: And to reiterate what @niemeyer and @morphis said, yes, the iw* tools aren't in the core snap so these rules don't have any effect. The snap that requires adminstering wireless networking should include these tools itself.

File: interfaces/builtin/wireless-control.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+network sna,
+
+network netlink,
+network netlink raw,
+network netlink dgram,
+
+@{PROC}/@{pid}/net/ r,
+@{PROC}/@{pid}/net/** r,
+
+/dev/wl* rw,
```
Comment (jdstrand) [inline interfaces/builtin/wireless-control.go]: I'm not familiar with this device. I thought all network interfaces were kernel virtual devices and didn't have device files in /dev. Can you comment?

File: interfaces/builtin/wireless-control.go [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+network packet,
+network pppox,
+network sna,
+
+network netlink,
+network netlink raw,
+network netlink dgram,
+
+@{PROC}/@{pid}/net/ r,
+@{PROC}/@{pid}/net/** r,
```
Comment (jdstrand) [inline interfaces/builtin/wireless-control.go]: FYI, everything above this comment is already in network-control with the exception of `network netlink,` and `network netwlink raw,`. Note, `network netlink,` covers `network netlink raw` and `network netlink dgram` so those wouldn't be needed.

Why do you need `network netlink,`?

File: interfaces/builtin/wireless-control.go [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+/dev/wl* rw,
+
+# wireless tools
+/{,usr/}{,s}bin/iw ixr,
+/{,usr/}{,s}bin/iwconfig ixr,
+/{,usr/}{,s}bin/iwevent ixr,
+/{,usr/}{,s}bin/iwgetid ixr,
+/{,usr/}{,s}bin/iiwlist ixr,
```
Comment (jdstrand) [inline interfaces/builtin/wireless-control.go]: Typo

**Comment by lpotter** [comment]
closing this, as it is not needed. Will add few things to other existing interfaces
